Adding grpc multiple stream support (-s) and max server side stream (-grpc-max-streams),...

ldemailly opened the pull request:
- `-s` (or `Streams` in GRPCRunner options) to multiply number of connections by given number of streams (multiplex)
- `-grpc-max-streams` option to the grpc ping server
- `-grpc-ping-delay` option to the ping client (and server) to delay the connection
- `-ping` to use ping instead of health as the grpc echo server (ie should be fortio on the backend)

Fixes #223; Manually tested number of sockets/connections (using lsof…) + unit tests

Also
- switches to go1.10.1 for build/release
- vendor update for 0.9
- Added UI for grpc-ping-delay and ping mode
- Improved JSON results with RunType (http/grpc differentiation and grpc mode details)

ldemailly committed:
Adding grpc multiple stream support (-s)

Fixes #223

Manually tested (using lsof…)
Need to add test(s)

ldemailly committed:
Remove numthread check as it can happen inside Run() anyway

Exercise the code in test (but doesn’t check for connections, did that
manually)

ldemailly committed:
Add `-grpc-max-streams` flag for server side

It doesn’t “fail” need to add more long lived streams

ldemailly committed:
Adding `-grpc-ping-delay` to ping client/server

ldemailly committed:
Switch ping client to return SI units (seconds) + add test + fix proto case

danehans commented on fgrpc/grpcrunner_test.go:
```diff
@@ -55,9 +55,53 @@ func TestGRPCRunner(t *testing.T) {
 }
 
+//TODO : find something that actually fails based on MAXSTREAM
+func TestGRPCRunnerMaxStreams(t *testing.T) {
```
It would be nice to see this as a table driven test.

ldemailly replied:
which table though ? there isn't much more than 1 thing to test

ldemailly committed:
New option to use `-ping` for grpc load tests and thus use `-grpc-ping-delay` to test max streams

ldemailly committed:
Remove todo

ldemailly committed:
Make linter happy, make test lest brittle

some of those tests are time sensitive, not really a way around that,
still better than no tests

ldemailly committed:
Remove duplication to make linter happy - at cost of weird code

ldemailly committed:
Try to reduce chance of failure on slow/overloaded ci (docker hub)

ldemailly committed:
Switch to v7 of build image with go1.10.1

ldemailly committed:
Vendor and version update

Lot’s of change in this PR; worth to be 0.9.0
(Including upcoming changes to json and UI)

ldemailly committed:
UI for ping and ping delay and better JSON (RunType added)

Adding UI for grpc-ping-delay and ping mode
Improved JSON with RunType

```
$ fortio load -grpc -payload abcdef -grpc-ping-delay 15us -ping -json -
localhost
Fortio 0.9.0-pre running at 8 queries per second, 8->8 procs, for 5s:
localhost
21:05:52 I grpcrunner.go:139> Starting GRPC Ping Delay=15µs
PayloadLength=6 test for localhost with 1*4 threads at 8.0 qps
Starting at 8 qps with 4 thread(s) [gomax 8] for 5s : 10 calls each
(total 40)
21:05:57 I periodic.go:533> T003 ended after 5.001385662s : 10 calls.
qps=1.9994458887621773
21:05:57 I periodic.go:533> T002 ended after 5.001357581s : 10 calls.
qps=1.9994571150020717
21:05:57 I periodic.go:533> T000 ended after 5.001426881s : 10 calls.
qps=1.9994294104326824
21:05:57 I periodic.go:533> T001 ended after 5.001529318s : 10 calls.
qps=1.9993884598478724
Ended after 5.001608538s : 40 calls. qps=7.9974
Sleep times : count 36 avg 0.55268342 +/- 0.002171 min 0.549344622 max
0.554996642 sum 19.8966031
Aggregated Function Time : count 40 avg 0.00073360315 +/- 0.0002007 min
0.000332043 max 0.001002607 sum 0.029344126
# range, mid point, percentile, count
>= 0.000332043 <= 0.001 , 0.000666021 , 97.50, 39
> 0.001 <= 0.00100261 , 0.0010013 , 100.00, 1
# target 50% 0.000666022
# target 75% 0.0008418
# target 90% 0.000947267
# target 99% 0.00100156
# target 99.9% 0.0010025
Ping SERVING : 40
All done 40 calls (plus 4 warmup) 0.734 ms avg, 8.0 qps
{
  "RunType": "GRPC Ping Delay=15µs PayloadLength=6",
  "Labels": "localhost , ldemailly-macbookpro",

```

ldemailly committed:
Update readme

ldemailly commented:
k it's ready to go now, PTAL,
preview image: docker.io `istio/fortio:0.9.0-pre1`

ldemailly commented:
it's also live on https://fortio-stage.istio.io/
(through https://github.com/istio/admin-sites/pull/63 )

danehans reviewed (commented):
I'm having a difficult time understanding how to use the new flags. I can only imagine what it's like for a new user. It would be helpful to add example usages to the readme.

I think it would also be helpful to organize the cli options using https://github.com/spf13/cobra. I know you want to limit deps, but cobra is widely used, well supported and will greatly improve cli usability.

danehans commented on README.md:
```diff
@@ -127,6 +132,8 @@ target is a url (http load tests) or host:port (grpc health test).  flags are:
 	List of pXX to calculate (default "50,75,90,99,99.9")
   -payload string
 	Payload string to send along
+  -ping
```
I thought the `load` command with the `-grpc` flag is responsible for performing a grpc load test?

ldemailly replied:
`fortio load -grpc` does health check grpc, `fortio load -grpc -ping` will now do ping instead of health grpc load

danehans replied:
ack

danehans commented:
What is the expected behavior if `-grpc-max-streams` is set to a smaller number than `-S` on the client? I expected to see an error or failure, but the test case worked:
```
$ docker run -d --ulimit nofile=20 --name fortio_server istio/fortio:0.9.0-pre1 server -grpc-max-streams 5 -loglevel debug

$ docker logs -f fortio_server
18:03:04 I logger.go:97> Log level is now 0 Debug (was 2 Info)
18:03:04 V stats.go:517> Will use [50 75 90 99 99.9] for percentiles
18:03:04 I pingsrv.go:66> Setting grpc.MaxConcurrentStreams server to 5
<SNIP>

$ docker exec fortio_server /usr/local/bin/fortio load -grpc -ping -s 10 -c 8 -grpc-ping-delay 2s localhost:8079
Fortio 0.9.0-pre running at 8 queries per second, 4->4 procs, for 5s: localhost:8079
18:23:49 I grpcrunner.go:139> Starting GRPC Ping Delay=2s test for localhost:8079 with 10*8 threads at 8.0 qps
```

ldemailly commented:
> I'm having a difficult time understanding how to use the new flags. I can only imagine what it's like for a new user. It would be helpful to add example usages to the readme.

yes better docs are always welcome :-)

> What is the expected behavior if -grpc-max-streams is set to a smaller number than -S on the client? I expected to see an error or failure, but the test case worked:

as you can see in the unit test, the behavior is just it slows down the additional streams, it's a go client feature to wait for a stream - there may/should be an option to grpc to tell it to fail instead of wait but I didn't search for that yet - can you ?

danehans commented:
@ldemailly why did the deps need to be upgraded to support this issue?

ldemailly commented:
> @ldemailly why did the deps need to be upgraded to support this issue?

it doesn't need to, I just want to rev up the vendor about once a month / as part of minor version bump and this PR has enough changes to warrant bumping to 0.9 (if we can get yours in too that'd be better)

danehans commented:
@ldemailly OK, I wanted to make sure I wasn't missing something in my review. I'm used to seeing separate dep bumps.

/lgtm

mandarjog reviewed (approved):
lgtm

ldemailly committed:
Code review feedback

Added an example and clarify -grpc flag
